From 51018af7a2104114ed9c00d9e80495d62e29e23b Mon Sep 17 00:00:00 2001
From: Josua Mayer <josua@solid-run.com>
Date: Fri, 21 Feb 2025 16:43:37 +0100
Subject: [PATCH 21/21] device property: add function to lookup a string from
 array by index

Sometimes drivers need to access individual elements from an array of
string. of.h already provides a suitable function which is specific to
device-tree.

Add a new function device_property_read_string_index similar to the
existing device_property_read_* functions that can look-up a specific
index inside a multiple string property.

Signed-off-by: Josua Mayer <josua@solid-run.com>
---
 drivers/base/property.c  | 55 ++++++++++++++++++++++++++++++++++++++++
 drivers/of/property.c    | 11 ++++++++
 include/linux/fwnode.h   |  5 ++++
 include/linux/property.h |  6 ++++-
 4 files changed, 76 insertions(+), 1 deletion(-)

diff --git a/drivers/base/property.c b/drivers/base/property.c
index 8c40abed7852..56e4d58b611b 100644
--- a/drivers/base/property.c
+++ b/drivers/base/property.c
@@ -231,6 +231,29 @@ int device_property_read_string(const struct device *dev, const char *propname,
 }
 EXPORT_SYMBOL_GPL(device_property_read_string);
 
+/**
+ * device_property_read_string_index - find a string in an array by index
+ * @dev: Device to get the property of
+ * @propname: Name of the property holding the array
+ * @index: The index of the string in the list of strings
+ * @val: The value is stored here
+ *
+ * Finds given index in a string array and stores the value into @val
+ * if found.
+ *
+ * Return: %0 if the string was found (success),
+ *	   %-EINVAL if given arguments are not valid,
+ *	   %-ENODATA if the property does not have a value,
+ *	   %-EPROTO or %-EILSEQ if the property type is not a string.
+ *	   %-ENXIO if no suitable firmware interface is present.
+ */
+int device_property_read_string_index(struct device *dev, const char *propname,
+				      int index, const char **val)
+{
+	return fwnode_property_read_string_index(dev_fwnode(dev), propname, index, val);
+}
+EXPORT_SYMBOL_GPL(device_property_read_string_index);
+
 /**
  * device_property_match_string - find a string in an array and return index
  * @dev: Device to get the property of
@@ -452,6 +475,38 @@ int fwnode_property_read_string(const struct fwnode_handle *fwnode,
 }
 EXPORT_SYMBOL_GPL(fwnode_property_read_string);
 
+/**
+ * fwnode_property_read_string_index - find a string in an array by index
+ * @fwnode: Firmware node to get the property of
+ * @propname: Name of the property holding the array
+ * @index: The index of the string in the list of strings
+ * @val: The value is stored here
+ *
+ * Read property @propname from the given firmware node and store the value into
+ * @val if found.  The value is checked to be a string.
+ *
+ * Return: %0 if the string was found (success),
+ *	   %-EINVAL if given arguments are not valid,
+ *	   %-ENODATA if the property does not have a value,
+ *	   %-EPROTO or %-EILSEQ if the property is not a string,
+ *	   %-ENXIO if no suitable firmware interface is present.
+ */
+int fwnode_property_read_string_index(const struct fwnode_handle *fwnode,
+				      const char *propname, int index,
+				      const char **val)
+{
+	int ret;
+
+	if (IS_ERR_OR_NULL(fwnode))
+		return -EINVAL;
+
+	ret = fwnode_call_int_op(fwnode, property_read_string_index, propname,
+				 index, val);
+
+	return ret < 0 ? ret : 0;
+}
+EXPORT_SYMBOL_GPL(fwnode_property_read_string_index);
+
 /**
  * fwnode_property_match_string - find a string in an array and return index
  * @fwnode: Firmware node to get the property of
diff --git a/drivers/of/property.c b/drivers/of/property.c
index 8044dfd18eb9..59854c56b961 100644
--- a/drivers/of/property.c
+++ b/drivers/of/property.c
@@ -935,6 +935,16 @@ of_fwnode_property_read_string_array(const struct fwnode_handle *fwnode,
 		of_property_count_strings(node, propname);
 }
 
+static int
+of_fwnode_property_read_string_index(const struct fwnode_handle *fwnode,
+				     const char *propname, int index,
+				     const char **val)
+{
+	const struct device_node *node = to_of_node(fwnode);
+
+	return of_property_read_string_index(node, propname, index, val);
+}
+
 static const char *of_fwnode_get_name(const struct fwnode_handle *fwnode)
 {
 	return kbasename(to_of_node(fwnode)->full_name);
@@ -1441,6 +1451,7 @@ const struct fwnode_operations of_fwnode_ops = {
 	.property_present = of_fwnode_property_present,
 	.property_read_int_array = of_fwnode_property_read_int_array,
 	.property_read_string_array = of_fwnode_property_read_string_array,
+	.property_read_string_index = of_fwnode_property_read_string_index,
 	.get_name = of_fwnode_get_name,
 	.get_name_prefix = of_fwnode_get_name_prefix,
 	.get_parent = of_fwnode_get_parent,
diff --git a/include/linux/fwnode.h b/include/linux/fwnode.h
index 5700451b300f..f1187d2b0db7 100644
--- a/include/linux/fwnode.h
+++ b/include/linux/fwnode.h
@@ -106,6 +106,7 @@ struct fwnode_reference_args {
  *			     success, a negative error code otherwise.
  * @property_read_string_array: Read an array of string properties. Return zero
  *				on success, a negative error code otherwise.
+ * @property_read_string_index: Find a string in an array by index.
  * @get_name: Return the name of an fwnode.
  * @get_name_prefix: Get a prefix for a node (for printing purposes).
  * @get_parent: Return the parent of an fwnode.
@@ -139,6 +140,10 @@ struct fwnode_operations {
 	(*property_read_string_array)(const struct fwnode_handle *fwnode_handle,
 				      const char *propname, const char **val,
 				      size_t nval);
+	int
+	(*property_read_string_index)(const struct fwnode_handle *fwnode_handle,
+				      const char *propname, int index,
+				      const char **val);
 	const char *(*get_name)(const struct fwnode_handle *fwnode);
 	const char *(*get_name_prefix)(const struct fwnode_handle *fwnode);
 	struct fwnode_handle *(*get_parent)(const struct fwnode_handle *fwnode);
diff --git a/include/linux/property.h b/include/linux/property.h
index d32b8052e086..76b6d1ffc1b9 100644
--- a/include/linux/property.h
+++ b/include/linux/property.h
@@ -55,7 +55,8 @@ int device_property_read_string(const struct device *dev, const char *propname,
 				const char **val);
 int device_property_match_string(const struct device *dev,
 				 const char *propname, const char *string);
-
+int device_property_read_string_index(struct device *dev, const char *propname,
+				      int index,const char **val);
 bool fwnode_property_present(const struct fwnode_handle *fwnode,
 			     const char *propname);
 int fwnode_property_read_u8_array(const struct fwnode_handle *fwnode,
@@ -75,6 +76,9 @@ int fwnode_property_read_string_array(const struct fwnode_handle *fwnode,
 				      size_t nval);
 int fwnode_property_read_string(const struct fwnode_handle *fwnode,
 				const char *propname, const char **val);
+int fwnode_property_read_string_index(const struct fwnode_handle *fwnode,
+				      const char *propname, int index,
+				      const char **val);
 int fwnode_property_match_string(const struct fwnode_handle *fwnode,
 				 const char *propname, const char *string);
 
-- 
2.43.0

